conformance/tests/gateway-infrastructure.go

```diff
@@ -0,0 +1,113 @@
+	ShortName:   "GatewayInfrastructure",
+	Description: "Propagation of metadata from Gateway infrastructure to generated components",
+	Features: []features.SupportedFeature{
+		features.SupportGateway,
+		features.SupportGatewayInfrastructurePropagation,
+	},
+	Manifests: []string{
+		"tests/gateway-infrastructure.yaml",
+	},
+	Test: func(t *testing.T, s *suite.ConformanceTestSuite) {
+		ns := "gateway-conformance-infra"
+
+		kubernetes.NamespacesMustBeReady(t, s.Client, s.TimeoutConfig, []string{ns})
+
+		gwNN := types.NamespacedName{
+			Name:      "gateway-with-infrastructure-metadata",
+			Namespace: "gateway-conformance-infra",
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), s.TimeoutConfig.DefaultTestTimeout)
+		defer cancel()
+
+		t.Logf("waiting for namespace %s and Gateway %s to be ready for testing", gwNN.Namespace, gwNN.Name)
+		kubernetes.GatewayMustHaveLatestConditions(t, s.Client, s.TimeoutConfig, gwNN)
+
+		t.Logf("retrieving Gateway %s/%s", gwNN.Namespace, gwNN.Name)
+		currentGW := &v1.Gateway{}
+		err := s.Client.Get(ctx, gwNN, currentGW)
+		require.NoError(t, err, "error getting Gateway: %v", err)
+		t.Logf("verifying that the Gateway %s/%s is accepted with infrastructure declared", gwNN.Namespace, gwNN.Name)
+		kubernetes.GatewayMustHaveCondition(t, s.Client, s.TimeoutConfig, gwNN, metav1.Condition{
+			Type:   string(v1.GatewayConditionAccepted),
+			Status: metav1.ConditionTrue,
+		})
+
+		t.Logf("verifying that generated resources for Gateway %s/%s have the proper gateway name label", gwNN.Namespace, gwNN.Name)
+		// Don't check services because implementations may have special filtering logic (e.g. for LB annotations)
+		// Instead, check service accounts for the gateway-name label first and then
+		// fallback to Pod if that fails
+		annotations := make(map[string]string, len(currentGW.Spec.Infrastructure.Annotations))
+		labels := make(map[string]string, len(currentGW.Spec.Infrastructure.Labels))
+		// Need to translate from (Annotation|Label)Key to string
+		for k, v := range currentGW.Spec.Infrastructure.Annotations {
+			annotations[string(k)] = string(v)
+		}
+
+		for k, v := range currentGW.Spec.Infrastructure.Labels {
+			labels[string(k)] = string(v)
+		}
+		saList := corev1.ServiceAccountList{}
+		podList := corev1.PodList{}
+		serviceList := corev1.ServiceList{}
+		err = s.Client.List(ctx, &saList, client.MatchingLabels{"gateway.networking.k8s.io/gateway-name": gwNN.Name}, client.InNamespace(ns))
+		require.NoError(t, err, "error listing ServiceAccounts")
+		err = s.Client.List(ctx, &podList, client.MatchingLabels{"gateway.networking.k8s.io/gateway-name": gwNN.Name}, client.InNamespace(ns))
+		require.NoError(t, err, "error listing Pods")
+		err = s.Client.List(ctx, &serviceList, client.MatchingLabels{"gateway.networking.k8s.io/gateway-name": gwNN.Name}, client.InNamespace(ns))
+		require.NoError(t, err, "error listing Services")
+		if len(saList.Items) > 0 {
```

Comment:
The structure if .. else ... else ... here implies that only one kind of resources is checked. Not sure if it's intentional, or do you want to check for all _applicable_ resources.

(As mentioned by Nick, Cilium implementation only creates  Service, hence above comment is mainly for other implementations.)

Reply:
Correct we only want to require that one of these resource types exists for a passing test

Reply:
Would it be more accurate to check that for _any_ resource with the matching label the annotations/labels have been applied, rather than "at least one", or is that expecting more than some implementations will want?

e.g. if a gateway creates both a labeled Deployment and Service should infra annotations and labels be populated on both?

Reply:
Hmm not a bad idea? Might be worth iterating on if we can get agreement on this first?

Reply:
That's actually what I'd meant by my [original comment](https://github.com/kubernetes-sigs/gateway-api/pull/3192#discussion_r1678525264), but @mikemorris explained it much better, +1 on this approach.

Reply:
Ah, sorry for misunderstanding! I went ahead and implemented that logic then 👍🏾

Reply:
Logic LGTM now!